diff --git a/tools/testing/selftests/bpf/prog_tests/d_path.c b/tools/testing/selftests/bpf/prog_tests/d_path.c                                                                                                                                                                                                                                                         18:08:55 [268/1874]
index ccc768592e66..1a2a2f1abf03 100644
--- a/tools/testing/selftests/bpf/prog_tests/d_path.c
+++ b/tools/testing/selftests/bpf/prog_tests/d_path.c
@@ -38,6 +38,14 @@ static int set_pathname(int fd, pid_t pid)
        return readlink(buf, src.paths[src.cnt++], MAX_PATH_LEN);
 }

+static inline long syscall_close(int fd)
+{
+       return syscall(__NR_close_range,
+                       (unsigned int)fd,
+                       (unsigned int)fd,
+                       0u);
+}
+
 static int trigger_fstat_events(pid_t pid)
 {
        int sockfd = -1, procfd = -1, devfd = -1;
@@ -104,18 +112,34 @@ static int trigger_fstat_events(pid_t pid)
        /* sys_close no longer triggers filp_close, but we can
         * call sys_close_range instead which still does
         */
-#define close(fd) syscall(__NR_close_range, fd, fd, 0)
+       syscall_close(pipefd[0]);
+       syscall_close(pipefd[1]);
+       syscall_close(sockfd);
+       syscall_close(procfd);
+       syscall_close(devfd);
+       syscall_close(localfd);
+       syscall_close(indicatorfd);
+       return ret;
+}

-       close(pipefd[0]);
-       close(pipefd[1]);
-       close(sockfd);
-       close(procfd);
-       close(devfd);
-       close(localfd);
-       close(indicatorfd);
+static void attach_and_load(struct test_d_path **skel)
+{
+       int err;

-#undef close
-       return ret;
+       *skel = test_d_path__open_and_load();
+       if (CHECK(!*skel, "setup", "d_path skeleton failed\n"))
+               goto cleanup;
+
+       err = test_d_path__attach(*skel);
+       if (CHECK(err, "setup", "attach failed: %d\n", err))
+               goto cleanup;
+
+       (*skel)->bss->my_pid = getpid();
+       return;
+
+cleanup:
+       test_d_path__destroy(*skel);
+       *skel = NULL;
 }

 static void test_d_path_basic(void)
@@ -124,16 +148,11 @@ static void test_d_path_basic(void)
        struct test_d_path *skel;
        int err;

-       skel = test_d_path__open_and_load();
-       if (CHECK(!skel, "setup", "d_path skeleton failed\n"))
-               goto cleanup;
-
-       err = test_d_path__attach(skel);
-       if (CHECK(err, "setup", "attach failed: %d\n", err))
+       attach_and_load(&skel);
+       if (!skel)
                goto cleanup;

        bss = skel->bss;
-       bss->my_pid = getpid();

        err = trigger_fstat_events(bss->my_pid);
        if (err < 0)
@@ -195,6 +214,39 @@ static void test_d_path_check_types(void)
        test_d_path_check_types__destroy(skel);
 }

+/* Check if the verifier correctly generates code for
+ * accessing the memory modified by d_path helper.
+ */
+static void test_d_path_mem_access(void)
+{
+       int localfd = -1;
+       char path_template[] = "/dev/shm/d_path_loadgen.XXXXXX";
+       struct test_d_path__bss *bss;
+       struct test_d_path *skel;
+
+       attach_and_load(&skel);
+       if (!skel)
+               goto cleanup;
+
+       bss = skel->bss;
+
+       localfd = mkstemp(path_template);
+       if (CHECK(localfd < 0, "trigger", "mkstemp failed\n"))
+               goto cleanup;
+
+       if (CHECK(fallocate(localfd, 0, 0, 1024) < 0, "trigger", "fallocate failed\n"))
+               goto cleanup;
+       remove(path_template);
+
+       if (CHECK(!bss->path_match_fallocate, "check",
+                 "failed to read fallocate path"))
+               goto cleanup;
+
+cleanup:
+       syscall_close(localfd);
+       test_d_path__destroy(skel);
+}
+
 void test_d_path(void)
 {
        if (test__start_subtest("basic"))
@@ -205,4 +257,7 @@ void test_d_path(void)

        if (test__start_subtest("check_alloc_mem"))
                test_d_path_check_types();
+
+       if (test__start_subtest("check_mem_access"))
+               test_d_path_mem_access();
 }
diff --git a/tools/testing/selftests/bpf/progs/test_d_path.c b/tools/testing/selftests/bpf/progs/test_d_path.c
index 84e1f883f97b..561b2f861808 100644
--- a/tools/testing/selftests/bpf/progs/test_d_path.c
+++ b/tools/testing/selftests/bpf/progs/test_d_path.c
@@ -17,6 +17,7 @@ int rets_close[MAX_FILES] = {};

 int called_stat = 0;
 int called_close = 0;
+int path_match_fallocate = 0;

 SEC("fentry/security_inode_getattr")
 int BPF_PROG(prog_stat, struct path *path, struct kstat *stat,
@@ -62,4 +63,26 @@ int BPF_PROG(prog_close, struct file *file, void *id)
        return 0;
 }

+SEC("fentry/vfs_fallocate")
+int BPF_PROG(prog_fallocate, struct file *file, int mode, loff_t offset, loff_t len)
+{
+       pid_t pid = bpf_get_current_pid_tgid() >> 32;
+       int ret = 0;
+       char path_fallocate[MAX_PATH_LEN] = {};
+
+       if (pid != my_pid)
+               return 0;
+
+       ret = bpf_d_path(&file->f_path,
+                        path_fallocate, MAX_PATH_LEN);
+       if (ret < 0)
+               return 0;
+
+       if (!path_fallocate[0])
+               return 0;
+
+       path_match_fallocate = 1;
+       return 0;
+}
+
 char _license[] SEC("license") = "GPL";
diff --git a/tools/testing/selftests/bpf/progs/verifier_bounds.c b/tools/testing/selftests/bpf/progs/verifier_bounds.c
index c49f44444752..b91056529f69 100644
--- a/tools/testing/selftests/bpf/progs/verifier_bounds.c
+++ b/tools/testing/selftests/bpf/progs/verifier_bounds.c
@@ -1709,6 +1709,160 @@ __naked void jeq_disagreeing_tnums(void *ctx)
        : __clobber_all);
 }

+SEC("socket")
+__description("conditional jump on same register, branch taken")
+__not_msg("20: (b7) r0 = 1 {{.*}} R0=1")
+__success __log_level(2)                                                                                                                                                                                                                                                                                                                                               18:08:55 [87/1874]
+__retval(0) __flag(BPF_F_TEST_REG_INVARIANTS)
+__naked void condition_jump_on_same_register(void *ctx)
+{
+       asm volatile("                  \
+       call %[bpf_get_prandom_u32];    \
+       w8 = 0x80000000;                \
+       r0 &= r8;                       \
+       if r0 == r0 goto +1;            \
+       goto l1_%=;                     \
+       if r0 >= r0 goto +1;            \
+       goto l1_%=;                     \
+       if r0 s>= r0 goto +1;           \
+       goto l1_%=;                     \
+       if r0 <= r0 goto +1;            \
+       goto l1_%=;                     \
+       if r0 s<= r0 goto +1;           \
+       goto l1_%=;                     \
+       if r0 != r0 goto l1_%=;         \
+       if r0 >  r0 goto l1_%=;         \
+       if r0 s> r0 goto l1_%=;         \
+       if r0 <  r0 goto l1_%=;         \
+       if r0 s< r0 goto l1_%=;         \
+l0_%=: r0 = 0;                         \
+       exit;                           \
+l1_%=: r0 = 1;                         \
+       exit;                           \
+"      :
+       : __imm(bpf_get_prandom_u32)
+       : __clobber_all);
+}
+
+SEC("socket")
+__description("jset on same register, constant value branch taken")
+__not_msg("7: (b7) r0 = 1 {{.*}} R0=1")
+__success __log_level(2)
+__retval(0) __flag(BPF_F_TEST_REG_INVARIANTS)
+__naked void jset_on_same_register_1(void *ctx)
+{
+       asm volatile("                  \
+       r0 = 0;                         \
+       if r0 & r0 goto l1_%=;          \
+       r0 = 1;                         \
+       if r0 & r0 goto +1;             \
+       goto l1_%=;                     \
+l0_%=: r0 = 0;                         \
+       exit;                           \
+l1_%=: r0 = 1;                         \
+       exit;                           \
+"      :
+       : __imm(bpf_get_prandom_u32)
+       : __clobber_all);
+}
+
+SEC("socket")
+__description("jset on same register, scalar value branch taken")
+__not_msg("12: (b7) r0 = 1 {{.*}} R0=1")
+__success __log_level(2)
+__retval(0) __flag(BPF_F_TEST_REG_INVARIANTS)
+__naked void jset_on_same_register_2(void *ctx)
+{
+       asm volatile("                  \
+       /* range [1;2] */               \
+       call %[bpf_get_prandom_u32];    \
+       r0 &= 0x1;                      \
+       r0 += 1;                        \
+       if r0 & r0 goto +1;             \
+       goto l1_%=;                     \
+       /* range [-2;-1] */             \
+       call %[bpf_get_prandom_u32];    \
+       r0 &= 0x1;                      \
+       r0 -= 2;                        \
+       if r0 & r0 goto +1;             \
+       goto l1_%=;                     \
+l0_%=: r0 = 0;                         \
+       exit;                           \
+l1_%=: r0 = 1;                         \
+       exit;                           \
+"      :
+       : __imm(bpf_get_prandom_u32)
+       : __clobber_all);
+}
+
+SEC("socket")
+__description("jset on same register, scalar value unknown branch 1")
+__msg("3: (b7) r0 = 0 {{.*}} R0=0")
+__msg("5: (b7) r0 = 1 {{.*}} R0=1")
+__success __log_level(2)
+__flag(BPF_F_TEST_REG_INVARIANTS)
+__naked void jset_on_same_register_3(void *ctx)
+{
+       asm volatile("                  \
+       /* range [0;1] */               \
+       call %[bpf_get_prandom_u32];    \
+       r0 &= 0x1;                      \
+       if r0 & r0 goto l1_%=;          \
+l0_%=: r0 = 0;                         \
+       exit;                           \
+l1_%=: r0 = 1;                         \
+       exit;                           \
+"      :
+       : __imm(bpf_get_prandom_u32)
+       : __clobber_all);
+}
+
+SEC("socket")
+__description("jset on same register, scalar value unknown branch 2")
+__msg("4: (b7) r0 = 0 {{.*}} R0=0")
+__msg("6: (b7) r0 = 1 {{.*}} R0=1")
+__success __log_level(2)
+__flag(BPF_F_TEST_REG_INVARIANTS)
+__naked void jset_on_same_register_4(void *ctx)
+{
+       asm volatile("                  \
+       /* range [-1;0] */              \
+       call %[bpf_get_prandom_u32];    \
+       r0 &= 0x1;                      \
+       r0 -= 1;                        \
+       if r0 & r0 goto l1_%=;          \
+l0_%=: r0 = 0;                         \
+       exit;                           \
+l1_%=: r0 = 1;                         \
+       exit;                           \
+"      :
+       : __imm(bpf_get_prandom_u32)
+       : __clobber_all);
+}
+
+SEC("socket")
+__description("jset on same register, scalar value unknown branch 3")
+__msg("4: (b7) r0 = 0 {{.*}} R0=0")
+__msg("6: (b7) r0 = 1 {{.*}} R0=1")
+__success __log_level(2)
+__flag(BPF_F_TEST_REG_INVARIANTS)
+__naked void jset_on_same_register_5(void *ctx)
+{
+       asm volatile("                  \
+       /* range [-1;1] */              \
+       call %[bpf_get_prandom_u32];    \
+       r0 &= 0x2;                      \
+       r0 -= 1;                        \
+       if r0 & r0 goto l1_%=;          \
+l0_%=: r0 = 0;                         \
+       exit;                           \
+l1_%=: r0 = 1;                         \
+       exit;                           \
+"      :
+       : __imm(bpf_get_prandom_u32)
+       : __clobber_all);
+}
+
 /* This test covers the bounds deduction when the u64 range and the tnum
  * overlap only at umax. After instruction 3, the ranges look as follows:
  *
